Document steps required to get Theano up-and-running on Ubuntu LTS (optimized)

mcfletch opened the pull request:
Includes full documentation on how to build an *optimized* Numpy/SciPy/Theano on current Ubuntu LTS (14.04), including the BLAS dependencies (and configuration). I've tested the code under Kubuntu 14.04 (Ubuntu derivative) for the git code path, but haven't actually tested the pip-install path. That said, I haven't actually tried with a clean install, I expect we are actually missing such things as build-essentials or the like, though that's not likely to be a problem for readers of the docs for Theano.

abergeron commented on doc/install_ubuntu.txt:
```diff
@@ -49,6 +49,35 @@ These instructions were written for Ubuntu 11.04, 11.10, 12.04, 12.10, 13.04,
+    # scipy dependency installation will fail
+    pip install numpy
+    
+    # Bleeding edge installation (Theano from git):
```
I would add more distinction to the two install blocks.  Like break out of the ::bash block and make the comment text on the page.  Otherwise it's too easy for people to think they have to somehow install theano twice.

mcfletch commented:
Okay, split out the bleeding edge version from the pip version, also added brief intro/link to the VirtualEnv docs in case user isn't familiar with the tool.

abergeron commented:
Some comment and questions I ask from others:

 - Why do you need to install libblas-dev?  This forces you to do the alternatives dance.
 - What is significantly different from the other Ubuntu instructions (apart from using virtualenv, which is not mandatory)?
 - Why are you installing scipy and numpy from the Ubuntu packages and then recompiling them inside the virtualenv?

mcfletch commented:
I was recompiling the libraries because I mis-read the first note on the page as implying that to get a *very* optimized version of Theano I could do "more work", which I took to mean compiling openblas-optimized versions of Numpy and SciPy. The inclusion of them in the apt-get was just to pull in any dependencies.  The inclusion of libblas-dev was just from the scipy-from-source build instructions, as was the "alternatives dance".

With that out of the way, there's no reason to do the building locally. I added 14.04 to the list of versions supported, simplified the list-of-versions to something easier to read, moved the confusing comment down to the performance testing section, added a VirtualEnv section, updated the bleeding-edge section to use an explicit code checkout, used code-blocks throughout, and added a comment about the seemingly outdated comment in the "manual openblas" section wrt thread support.

Feel free to use or not use those changes as you see fit.

abergeron commented on doc/install_ubuntu.txt:
```diff
@@ -144,13 +164,22 @@ system package, you can run this:
+.. code-block: bash
+    
+    git pull
+    python setup.py develop --user
```
Now there are two "Bleeding edge" sections.  Could you make it clearer that this is to "Update Bleeding Edge" (perhaps by changing the section title to just that).

Also to update a checkout you only need the `git pull` step, the setup.py call is needed once.

abergeron commented on doc/install_ubuntu.txt:
```diff
@@ -144,13 +164,22 @@ system package, you can run this:
+
+    I believe this is outdated, my machine seems to be using 8 threads 
+    happily with the binary openblas...
+
```
This means that we can probably delete this section, but I haven't personally encountered this so I'll wait a bit to see if somebody else comments, otherwise, do as you wish.

nouiz replied:
With ubuntu 12.04, on my laptop with 4 core, there is only 2 cores used. So this could depend of the version of the OS.

abergeron commented:
All right, I like this direction better.  I just have a small nit about updating and I'm good with merging.

mcfletch commented:
Okay, altered as requested. Note that the setup.py call is required in packages when there is a new command-line entry point (using setuptools entry points). I've removed it, as I'm not seeing any setuptools entry points in the setup script (you seem to be using `scripts` instead).

One more note: at some point you should likely drop the 11.04 instructions to simplify the page. 11.04 was end-of-life'd in 2012-10. While 11.10, 12.10, 13.04 and 13.10 are also EOL'd their instructions are the same as for current, so they don't complicate the documentation.

HTH

abergeron commented:
It's good for me.  I'll just way a bit for the merge to give other people the time to comment.

abergeron commented:
It seems everybody is either fine with this or indifferent.  So I'll merge it.
